src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,165 @@
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node newNode = new Node(null, value, null);
+        if (isEmpty()) {
+            last = newNode;
+            first = newNode;
+        } else {
+            newNode.prev = last;
+            last.next = newNode;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size);
+        Node newNode = new Node(null, value, null);
+        if (first == null || index == size) {
+            add(value);
+        } else if (index == 0) {
+            newNode.next = first;
+            first.prev = newNode;
+            first = newNode;
+            size++;
+        } else {
+            Node<T> current = getNodeByIndex(index - 1);
+            newNode.next = current.next;
+            current.next.prev = newNode;
+            current.next = newNode;
+            newNode.prev = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size - 1);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size - 1);
+
+        T removedElement = null;
+        if (index == 0) {
+            removedElement = removeFirst();
+        } else if (index == size - 1) {
+            removedElement = removeLast();
+        } else {
+            removedElement = getNodeByIndex(index).item;
+            removeNode(getNodeByIndex(index));
+        }
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        while (current != null) {
+            if (current.item == null ? object == null : current.item.equals(object)) {
+                removeNode(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return first == null;
+    }
+
+    private void checkIndex(int index, int size) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current = first;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
```

Comment:
https://github.com/mate-academy/jv-linked-list/blob/master/checklist.md#dont-iterate-your-nodes-from-the-head-to-the-tail-or-vice-versa